android-components/components/browser/engine-gecko/src/main/java/mozilla/components/experiment/NimbusExperimentDelegate.kt

```diff
@@ -0,0 +1,99 @@
+import org.mozilla.geckoview.GeckoResult
+
+/**
+ * Default Nimbus [ExperimentDelegate] implementation to communicate with mobile Gecko and GeckoView.
+ */
+class NimbusExperimentDelegate : ExperimentDelegate {
+
+    private val logger = Logger(NimbusExperimentDelegate::javaClass.name)
+
+    /**
+     * Retrieves experiment information on the feature for use in GeckoView.
+     *
+     * @param feature Nimbus feature to retrieve information about
+     * @return a [GeckoResult] with a JSON object containing experiment information or completes exceptionally.
+     */
+    override fun onGetExperimentFeature(feature: String): GeckoResult<JSONObject> {
+        val result = GeckoResult<JSONObject>()
+        val nimbusFeature = GeckoNimbus.getFeature(feature)
+        if (nimbusFeature != null) {
+            result.complete(nimbusFeature.toJSONObject())
+        } else {
+            logger.warn("Could not find Nimbus feature '$feature' to retrieve experiment information.")
+            result.completeExceptionally(ExperimentException(ERROR_FEATURE_NOT_FOUND))
+        }
+        return result
+    }
+
+    /**
+     * Records that an exposure event occurred with the feature.
+     *
+     * @param feature Nimbus feature to record information about
+     * @return a [GeckoResult] that completes if the feature was found and recorded or completes exceptionally.
+     */
+    override fun onRecordExposureEvent(feature: String): GeckoResult<Void> {
+        val result = GeckoResult<Void>()
+        val nimbusFeature = GeckoNimbus.getFeature(feature)
+        if (nimbusFeature != null) {
+            nimbusFeature.recordExposure()
+            result.complete(null)
+        } else {
+            logger.warn("Could not find Nimbus feature '$feature' to record an exposure event.")
+            result.completeExceptionally(ExperimentException(ERROR_FEATURE_NOT_FOUND))
+        }
+        return result
+    }
+
+    /**
+     * Records that an exposure event occurred with the feature, in a given experiment.
+     * Note: See [onRecordExposureEvent] if no slug is known or needed
+     *
+     * @param feature Nimbus feature to record information about
+     * @param slug Nimbus experiment slug to record information about
+     * @return a [GeckoResult] that completes if the feature was found and recorded or completes exceptionally.
+     */
+    override fun onRecordExperimentExposureEvent(feature: String, slug: String): GeckoResult<Void> {
+        val result = GeckoResult<Void>()
+        val nimbusFeature = GeckoNimbus.getFeature(feature)
+        if (nimbusFeature != null) {
+            nimbusFeature.recordExperimentExposure(slug)
+            result.complete(null)
+        } else {
+            logger.warn("Could not find Nimbus feature '$feature' to record an exposure event.")
+            result.completeExceptionally(ExperimentException(ERROR_FEATURE_NOT_FOUND))
+        }
+        return result
+    }
```

Comment:
Nit: is it a premature generalization to factor out the common code— you can repeat for `recordExposure` and `recordMalformedConfiguration`?

```suggestion
    override fun onRecordExperimentExposureEvent(feature: String, slug: String): GeckoResult<Void> {
        withFeature(feature) { it.recordExperimentExposure(slug) }
    }
    
    fun withFeature(featureId: String, closure: (FeatureHolder<*>) -> ()) {
        val result = GeckoResult<Void>()
        val nimbusFeature = GeckoNimbus.getFeature(feature)
        if (nimbusFeature != null) {
            closure(nimbusFeature)
            result.complete(null)
        } else {
            logger.warn("Could not find Nimbus feature '$feature' to record an exposure event.")
            result.completeExceptionally(ExperimentException(ERROR_FEATURE_NOT_FOUND))
        }
        return result
    }
```

Reply:
Nice idea and looks much neater, updated!